feat: Add training progression tracking feature for experimental impl…

abhijeet-dhumal opened the pull request:
<!--  Thanks for sending a pull request! Here are some tips for you:
1. If this is your first time, check our contributor guidelines: https://www.kubeflow.org/docs/about/contributing
2. To know more about Kubeflow SDK, check the developer guide:
    https://github.com/kubeflow/sdk/blob/main/CONTRIBUTING.md
3. If you want *faster* PR reviews, check how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
-->

**What this PR does / why we need it**:
[RHOAIENG-38269](https://issues.redhat.com/browse/RHOAIENG-38269)
Add experimental support for HuggingFace Transformers training jobs with TransformersTrainer (built-in instrumentation and progression tracking capabilities) under RHAITrainer types
Trainer controller side related instrumentation in review here : https://github.com/opendatahub-io/trainer/pull/20 🏁 

Default progression metrics format used by TransformersTrainer for enabled progression tracking : 
Example demo tested & verified ✅  : https://github.com/abhijeet-dhumal/sdk/blob/add-progression-example/examples/transformers-text-classification.ipynb
```
{
        "progressPercentage": 37,
        "estimatedRemainingSeconds": 217,
        "currentStep": 283,
        "totalSteps": 750,
        "currentEpoch": 1,
        "totalEpochs": 2,
        "trainMetrics": {
          "loss": 0.0352,
          "learning_rate": 0.000031400000000000004,
          "grad_norm": 0.0012,
          "throughput_samples_sec": 3.8
        },
        "evalMetrics": {
          "eval_loss": 0.0421,
          "eval_accuracy": 0.89
        }
 }
```
<img width="2494" height="1594" alt="Screenshot 2025-11-20 at 3 43 41 PM" src="https://github.com/user-attachments/assets/b75a9bdc-dbfa-499b-9d84-14b14301694c" />

<img width="2486" height="1448" alt="Screenshot 2025-11-20 at 3 37 54 PM" src="https://github.com/user-attachments/assets/11f00f27-0bdd-4c26-a618-c0557e927610" />

**Checklist:**

- tested scenarios : 
   - The base distributed Training support using Kubeflow Trainer operator : https://github.com/opendatahub-io/trainer/tree/main - `make deploy-rhoai` ✅ 
   - Trainer Controller side instrumentation for progression tracking support : https://github.com/opendatahub-io/trainer/pull/20 ✅ 
   - The Transformers trainer, TRL (SFTTrainer) support ✅ 
   - The Muti-node Multi-GPU (CPU-Gloo, GPU- NVIDIA (nccl)) support for training ✅ 


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Public RHAI trainer alias to unify trainer types.
  * Transformers trainer: optional progression tracking with configurable metrics port and poll interval, input validation, and automatic instrumentation injected into training jobs.
  * New standard annotation keys and a utility to merge progression/metrics annotations into job metadata.
* **Bug Fixes**
  * Clearer error handling for unknown/unsupported trainer types (now raises an explicit error).
* **Tests**
  * Extensive unit tests for progression tracking, instrumentation wrapper generation, validation, and job wiring.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

robert-bell commented:
Thanks for raising this @abhijeet-dhumal! The implementation is quite different from the one I sketched out in https://github.com/robert-bell/kubeflow-sdk/pull/1, e.g. you've got an ExperimentalKubernetesBackend, the `TransformersTrainer` inherits from `CustomTrainer`. Was there a reason for this? If not, could we use the approach from that sketch please as it's what we'd agreed with the devx team? I think @briangallagher is working on a similar PR for training hub using the approach from the sketch.

Could we also split out the checkpointing work for a future PR please? Let's try keep this change smaller for now so it's easier to test and check it's working.

abhijeet-dhumal commented:
> Thanks for raising this @abhijeet-dhumal! The implementation is quite different from the one I sketched out in [robert-bell#1](https://github.com/robert-bell/kubeflow-sdk/pull/1), e.g. you've got an ExperimentalKubernetesBackend, the `TransformersTrainer` inherits from `CustomTrainer`. Was there a reason for this? If not, could we use the approach from that sketch please as it's what we'd agreed with the devx team? I think @briangallagher is working on a similar PR for training hub using the approach from the sketch.
> 
> Could we also split out the checkpointing work for a future PR please? Let's try keep this change smaller for now so it's easier to test and check it's working.

Hi @robert-bell
I have tried addressing both concerns:
TransformersTrainer is now standalone and no longer inherits from CustomTrainer.

I have updated the quickstart example in this PR description an have already tested using controller side instrumentation here : https://github.com/opendatahub-io/trainer/pull/17/commits/6be6a808bdd695a29c68b7db24580399e4bbb08e
<img width="1450" height="789" alt="Screenshot 2025-11-11 at 2 05 14 PM" src="https://github.com/user-attachments/assets/80a3af5c-56b1-41b6-9735-7e9999c54668" />


Could you please check once, I want to ensure the ExperimentalKubernetesBackend approach aligns with what was agreed. Happy to refactor if needed.

Now action items remaining : 

- Naming convention for the experimental prefix.. 

- TRL (Training Reinforcement Learning) support via TransformersTrainer

robert-bell commented:
Hey @abhijeet-dhumal, could you please check out https://github.com/opendatahub-io/kubeflow-sdk/pull/22 that @briangallagher is working on? We need to align with the approach in his PR so we can rebase on top off it. 

In particular, I don't think we need the `ExperimentalKubernetesBackend`, and I'm hoping that most/all of the changes can be contained within `kubeflow/trainer/rhai/transformers.py`.

(nb - the current plan is to avoid the word "experimental" to give customers more confidence that this is GA-ready, so we're renaming `experimental` to `rhai`.)

robert-bell commented on kubeflow/trainer/experimental/instrumentation/callbacks.py:
```diff
@@ -0,0 +1,140 @@
+from typing import Optional
+
+from kubeflow.trainer.experimental.instrumentation.http_server import get_http_server_code
+from kubeflow.trainer.experimental.instrumentation.templates import get_progress_callback_code
```
issue: at runtime we're not going to have the sdk installed so we're not going to be able to import any custom code. Would it be easier to put all this code into a single file and inject both the server and callback together?

robert-bell commented on kubeflow/trainer/experimental/instrumentation/templates/http_server.py:
```diff
@@ -0,0 +1,128 @@
+                    else:
+                        target[key] = value
+
+            deep_update(cls.metrics, updates)
```
suggestion: the contract only supports float metric values, are we able to avoid needing the recursive update?

robert-bell commented on kubeflow/trainer/experimental/instrumentation/templates/progress_callback.py:
```diff
@@ -0,0 +1,210 @@
+        elif minutes > 0:
+            return f"{minutes}m {secs}s"
+        else:
+            return f"{secs}s"
```
question: is this for making the human-readable `estimatedRemainingTimeSummary` field? If it is, I'd suggest we let the KFT controller make this from the `estimatedReaminingSeconds` field. That way it only needs doing once centrally in the control plane :raised_hands:

abhijeet-dhumal replied:
I have updated the implementation to keep the SDK metrics format concise which further can be handled by controller for more insightful fields by keeping the instrumentation wrappers less crowded. ✅ 
Please review !

robert-bell commented on kubeflow/trainer/backends/kubernetes/backend.py:
```diff
@@ -176,7 +176,7 @@ def train(
         initializer: Optional[types.Initializer] = None,
         trainer: Optional[
-            Union[types.CustomTrainer, types.CustomTrainerContainer, types.BuiltinTrainer]
+            Union[types.CustomTrainer, types.CustomTrainerContainer, types.BuiltinTrainer, Any]
```
suggestion: rather than `Any` can we use `RHAITrainer` here?

robert-bell commented on kubeflow/trainer/backends/kubernetes/backend.py:
```diff
@@ -593,7 +622,7 @@ def _get_trainjob_spec(
         initializer: Optional[types.Initializer] = None,
         trainer: Optional[
-            Union[types.CustomTrainer, types.CustomTrainerContainer, types.BuiltinTrainer]
+            Union[types.CustomTrainer, types.CustomTrainerContainer, types.BuiltinTrainer, Any]
```
suggestion: rather than `Any` can we use `RHAITrainer` here?

robert-bell commented on kubeflow/trainer/backends/kubernetes/backend.py:
```diff
@@ -209,6 +209,35 @@ def train(
+            if spec_annotations is None:
+                spec_annotations = progression_annotations
+            else:
+                spec_annotations = {**spec_annotations, **progression_annotations}
```
suggestion: can we move this logic into `rhai/utils.py` to be the same as `get_trainer_crd_from_transformers_trainer`?

Ideally we want the changes in upstream files to be as minimal as possible.

How about tidying the if clause a little?

```suggestion
        if isinstance(trainer, RHAITrainer):
            rhai_annotations = rhai_utils.get_rhai_train_job_annotations(trainer)
            if spec_annotations is None:
                spec_annotations = rhai_annotations
            else:
                # apply spec_annotations last to allow users to override the rhai annotations 
                spec_annotations = rhai_annotations | spec_annotations
```

robert-bell replied:
Did you mean to apply another change here? This is still a big blob of extra code in the upstream `backend.py` file which is likely to cause bigger merge conflicts.

robert-bell commented on kubeflow/trainer/rhai/__init__.py:
```diff
@@ -0,0 +1,35 @@
+
+This module provides RHAI trainer types and utilities:
+- TransformersTrainer: HuggingFace Transformers/TRL with instrumentation
+- TrainingHubTrainer: (future) RHAI Training Hub integration
```
nit: make sure to update this comment when you rebase on #22

robert-bell commented on kubeflow/trainer/rhai/constants.py:
```diff
@@ -0,0 +1,23 @@
+ANNOTATION_PROGRESSION_TRACKING = "trainer.opendatahub.io/progression-tracking"
+ANNOTATION_METRICS_PORT = "trainer.opendatahub.io/metrics-port"
+ANNOTATION_METRICS_POLL_INTERVAL = "trainer.opendatahub.io/metrics-poll-interval"
+ANNOTATION_FRAMEWORK = "trainer.opendatahub.io/framework"
```
nit: do we need the framework annotation? Is it being used in the backend? If not, I'd suggest we keep things simple and remove it. wdyt?

abhijeet-dhumal replied:
IMHO framework makes more sense as we are supporting the Transformers and TrainingHub via as part of RHAI trainers.. so it will make more sense for user to identify which framework this TrainJob is using or belongs too 
WDYT?

robert-bell replied:
I don't have a strong opinion on it. I'd argue given the target data scientist persona is unlikely they'll see the annotation, but it doesn't present much risk.

abhijeet-dhumal replied:
IMO as we will be supporting the Transformers and TrainerHub frameworks as part of productised SDK, to categorise the TrainJobs based on framework agnostic searching via annotation makes sense..
I think it will also help in analytics part to discover which framework specific trainjobs are being used by customers in managed environments..

```
oc get trainjobs -o json | jq -r '.items[] | select(.metadata.annotations["trainer.opendatahub.io/framework"] == "transformers")'
oc get trainjobs -o json | jq -r '.items[] | select(.metadata.annotations["trainer.opendatahub.io/framework"] == "traininghub")'
```

robert-bell replied:
I'm happy to keep this in then.

briangallagher replied:
No objection here just an observation.
A framework label is used in the runtime also - [see PR](https://github.com/opendatahub-io/trainer/pull/21/files#diff-3493ffed0f3b1b96f35310712fedfc2363ec5bfaa0db20fc3cfa75547702da9e) 

For the traininghub runtime it's `trainer.kubeflow.org/framework: training-hub` I'll update that to **traininghub** to align it. 

For transformers I guess the runtime will be the torch distributed? So, the runtime framework will be **_torch_** and trainjob framework will be **_transformers_**? I don't think this is a problem, just highlighting.

abhijeet-dhumal replied:
Thanks @briangallagher for sharing these insights! 👀

robert-bell commented on kubeflow/trainer/rhai/utils.py:
```diff
@@ -0,0 +1,297 @@
+# =============================================================================
+
+
+def get_job_progress(
```
nit: I suggest we hold off implementing this just yet. We can add it in a future PR. It'd maybe make more sense to be part of the `TrainJob` response model, but let's not update that code yet until we have better consensus upstream.

abhijeet-dhumal replied:
Actually the get_job_progress method is closely tied to progression tracking if we see from customer perspective..
While testing this feature using notebook, this method was single source of truth so I would be inclined towards keeping this :)

robert-bell replied:
Let's pull this into a follow on PR. It needs a bit more thought on what the API should be - ideally it should be a property of `TrainJob` object itself but that's going to cause issues with upstream compat.

robert-bell commented on kubeflow/trainer/rhai/README.md:
question: is this the right place to document this? @briangallagher what was your plan?

Perhaps we can hold off adding these docs to this PR and add them to a follow on PR.

robert-bell commented on kubeflow/trainer/rhai/transformers.py:
```diff
@@ -0,0 +1,407 @@
+            self.send_response(404)
+            self.send_header("Content-Type", "text/plain")
+            self.end_headers()
+            self.wfile.write(b"Not Found")
```
suggestion: we don't need to support healthz, so I think we can just ignore the path completely

```suggestion
       self.send_response(200)
       self.send_header("Content-Type", "application/json")
       self.end_headers()
       metrics_json = ...
       self.wfile.write(metrics_json.encode("utf-8"))
```

abhijeet-dhumal replied:
IMO the explicit checks will be needed for debugging purpose incase any issue comes in
I'd be happy to discuss if I've misunderstood you!

robert-bell commented on kubeflow/trainer/rhai/transformers.py:
```diff
@@ -0,0 +1,407 @@
+
+    custom_metrics_json = json.dumps(custom_metrics)
+
+    return f'''
```
suggestion: what about putting this code in a function (i.e. not a string) and then turning it into a string using `inspect.getsource`? This would give us syntax highlighting for better devx and also let us test the code more easily.

```python

def create_instrumentation(metrics_port: int, ...other_args):
    import http.server
    import 
    from transformers import Trainer, TrainerCallback
    
    class KubeflowProgressCallback(TrainerCallback):
        ...
    
    class ProgressMetricsHandler(http.server.BaseHTTPRequestHandler):
        ...

    def apply_instrumentation():
        original_init = Trainer.__init__
        def instrumented_init(self, *args, **kwargs):
            ...
        Trainer.__init__ = instrumented_init

   # return everything for testing (maybe?)
   return apply_instrumentation, KubeflowProrgessCallback, ProgressMetricsHandler
        

def get_command(trainer: TransformersTrainer) -> str:
    instrumentation_code = inspect.getsource(create_instrumentation)
    instrumentation_code = textwrap.dedent(instrumentation_code)

   user_code = inspect.getsource(trainer.func)
   user_call  # ...all the other bits that the sdk currently does for the CustomTrainer
    
    return textwrap.dedent(f"""
        {instrumentation_code}
        apply_instrumentation, *_ = {create_instrumentation.__name__}(metrics_port={trainer.metrics_port})
        apply_instrumentation()

        {user_code}
        {user_call}
    """
```

That code snippet doesn't include installing python packages - that still needs working out. wdyt? Happy to touch base on a call if it's not clear.

robert-bell commented on kubeflow/trainer/rhai/transformers.py:
```diff
@@ -0,0 +1,407 @@
+class ProgressMetricsHandler(http.server.BaseHTTPRequestHandler):
+    """HTTP handler that serves training progress metrics as JSON."""
+
+    metrics = {{
```
nit: I think you mentioned it, but the metrics fields needs updating to match the spec

robert-bell replied:
issue: I'm wary of making this a class variable (global state).

If we use the function approach from https://github.com/opendatahub-io/kubeflow-sdk/pull/21/files#r2524129705 could we instead make this a nonlocal variable in the `create_instrumentation` function?

robert-bell commented on kubeflow/trainer/rhai/transformers.py:
```diff
@@ -0,0 +1,407 @@
+            if self.custom_metrics_map:
+                for log_key, metric_name in self.custom_metrics_map.items():
+                    if log_key in logs:
+                        metrics_update[metric_name] = logs[log_key]
```
question: what does transformers put in `logs`? Is it just what the user has configured via the `Trainer` `compute_metrics` argument? If so, is there any reason not to add every item from `logs` and remove the `TransformersTrainer.custom_metrics` field completely? If the user doesn't want a metric logged, they can just remove it from their Trainer? wdyt?

abhijeet-dhumal replied:
Actually Transformers supports wide range of metrics and user can extend those via custom params to be accessed in logs, the custom_metrics field allows users to track additional metrics beyond the standard ones we automatically capture (like loss, learning_rate, and eval metrics). When users log custom domain-specific metrics in their Transformers training code—such as perplexity, BLEU scores, or business KPIs—these appear in the logs but won't be automatically tracked by our default filters. By mapping these metrics in custom_metrics, users can explicitly tell the progression tracker which additional logged metrics should be included in progression tracking, and they can control whether each metric is categorized as a training or evaluation metric based on the mapped name's prefix suported by transformers. The field defaults to an empty dict so it's zero-config for common cases, but gives power users a clear way to extend tracking to their specific metrics without needing to modify the core metric-tracking logic.
I'd be happy to discuss if I've misunderstood you!

robert-bell commented on kubeflow/trainer/rhai/transformers.py:
```diff
@@ -0,0 +1,407 @@
+            if self.custom_metrics_map:
+                for log_key, metric_name in self.custom_metrics_map.items():
+                    if log_key in logs:
+                        metrics_update[metric_name] = logs[log_key]
```
question: can you check transformers always provide us with float values? Looking at, e.g. [transformers mlflow integration](https://github.com/huggingface/transformers/blob/v4.57.1/src/transformers/integrations/integration_utils.py#L1497), the value could be a pytorch value.

Maybe we should do something similar
 
```suggestion
                        v = logs[log_key]
                        if isinstance(v, (int, float)):
                            metrics_update[metric_name] = v
                        elif isinstance(v, torch.Tensor) and v.numel() == 1:
                            metrics_update[metric_name] = v.item()
                        else:
                            logger.warning(...some warning message)
```

abhijeet-dhumal replied:
Oh, great catch ! 🙌  👀

robert-bell commented on kubeflow/trainer/rhai/transformers.py:
```diff
@@ -0,0 +1,407 @@
+        status_msg = (
+            f"Training in progress: {{progress_pct:.1f}}% complete, "
+            f"{{remaining_str}} remaining"
+        )
```
nit: I think we don't need this. Let's let the control plane format this message from the estimated remaining seconds

robert-bell commented on kubeflow/trainer/rhai/transformers.py:
```diff
@@ -0,0 +1,407 @@
+    server_thread.start()
+
+    print(f"[Kubeflow] Metrics server started on port {{port}}", flush=True)
+    return server
```
suggestion: maybe the KubeflowProgressCallback should create and start the server, possibly the same way the MLFlow Transformers [callback works](https://github.com/huggingface/transformers/blob/8cb5963cc22174954e7dca2c0a3320b7dc2f4edc/src/transformers/integrations/integration_utils.py#L1493).

We should maybe be starting the server on the rank-0 process too, which would be relatively easy using the Transformer callback's `state.is_world_process_zero` variable.

robert-bell commented on kubeflow/trainer/rhai/utils.py:
```diff
@@ -0,0 +1,297 @@
+        )
+
+
+def get_trainer_crd_from_transformers_trainer(
```
nit: I think this should go in the `transformers.py` file to keep all the transformers code together. There might be some shared utility bits that can go here (e.g. converting user func into a string, or the pypi import bits).

wdty?

robert-bell commented on kubeflow/trainer/rhai/transformers_test.py:
```diff
@@ -0,0 +1,456 @@
+    assert "{{user_func_import_and_call}}" in wrapper
+
+    # Verify port is injected correctly
+    assert f"port={28080}" in wrapper or f"port = {28080}" in wrapper
```
suggestion: I'm not sure how useful looking at the stringified instrumentation code is in the tests. If we go with the suggestion of putting this instrumentation code into a function, we can call that function in the tests and actually test the TrainerCallback and the http server behaviour properly. We can mock the transformers library rather than add it as a project dependency. wdyt?

abhijeet-dhumal committed:
feat: Add training progression tracking feature for experimental TransformerTraininer implementation

Signed-off-by: abhijeet-dhumal <abhijeetdhumal652@gmail.com>

abhijeet-dhumal commented:
@Fiona-Waters @szaher @briangallagher @kramaranya May I request DevX team to review this PR now please 🙌 
Thanks a lot in advance !!

robert-bell commented on kubeflow/trainer/rhai/transformers.py:
```diff
@@ -22,15 +74,204 @@ class TransformersTrainer:
+    custom_metrics: dict[str, str] = field(default_factory=dict)
+
+
+def get_transformers_instrumentation_wrapper(
```
there's a lot of code being constructed into a string here which may be error-prone and is hard to test and maintain.

In particular, `_metrics_state` is defined both here inside a string and in `progression_instrumentation.py` -- these need keeping in sync.

I would suggest we consider an implementation closer to what I proposed previously (https://github.com/opendatahub-io/kubeflow-sdk/pull/21#discussion_r2524129705) -- a single function that contains everything and returns a function that does the instrumentation.

Happy to touch base about it if you want my input.

abhijeet-dhumal replied:
I have refactored the implementation details accordingly, please review!

robert-bell replied:
lgtm - thanks for making this change 🙌

robert-bell commented on kubeflow/trainer/rhai/transformers.py:
```diff
@@ -10,8 +25,45 @@
+                                       higher controller load. Recommended range: 5-60 seconds.
+        custom_metrics: Dict mapping log keys to metric names to track.
+                       Example: {"eval_accuracy": "accuracy", "custom_metric": "custom"}
+                       These will be exposed in the /metrics endpoint.
```
question: is the aim of this to let users "rename" the metrics so they can pick which appear under "evalMetrics" and which ones under "trainMetrics"? Does the user still need to actually "add" the metrics inside their `train_fn`?

If it is, I think this is a bit confusing. What about either saying users should prefix all eval metrics with `eval_` or calling this field `eval_metrics: list[str] | None` and any metric that's listed here appears under `evalMetrics`, otherwise it gets put under `trainMetric`?

Let me know if I've misunderstood something - we should chat about it over a call.

abhijeet-dhumal replied:
As we discussed this already in the call, let's make the custom metrics implementation out of the loop for now and can be added in the follow up PR 🙌

robert-bell commented on kubeflow/trainer/rhai/progression_instrumentation.py:
```diff
@@ -0,0 +1,236 @@
+
+
+# Module-level state (encapsulated, not global class state)
+_metrics_state: dict[str, Any] = {
```
suggestion: can we make this a dataclass to avoid typos in the fields.

robert-bell commented on kubeflow/trainer/rhai/progression_instrumentation.py:
```diff
@@ -0,0 +1,236 @@
+            self.wfile.write(_get_metrics_json().encode("utf-8"))
+        else:
+            self.send_response(404)
+            self.end_headers()
```
I appreciate you marked this comment resolved previously but I'll re-raise it here to allow the devx team to have visibility https://github.com/opendatahub-io/kubeflow-sdk/pull/21#discussion_r2523994559:

> suggestion: we don't need to support healthz, so I think we can just ignore the path completely

abhijeet-dhumal replied:
I have removed this part, please review!

robert-bell replied:
lgtm - thanks Abhijeet!

abhijeet-dhumal committed:
fix: Refator instrumentation code

Signed-off-by: abhijeet-dhumal <abhijeetdhumal652@gmail.com>

abhijeet-dhumal committed:
fix: Remove get_job_progress and custom_metrics supports and refactor components

Signed-off-by: abhijeet-dhumal <abhijeetdhumal652@gmail.com>

abhijeet-dhumal committed:
fix: use 3.9 compatible approach

Signed-off-by: abhijeet-dhumal <abhijeetdhumal652@gmail.com>

efazal commented on kubeflow/trainer/rhai/transformers.py:
```diff
@@ -22,15 +54,328 @@ class TransformersTrainer:
+        def _instrumented_trainer_init(self, *args, **kwargs):
+            result = _original_init(self, *args, **kwargs)
+            callback = KubeflowProgressCallback(metrics_port)
+            if callback not in self.callback_handler.callbacks:
```
Maybe we should do a callback type check here rather than instance check to avoid the KubeflowProgressCallback being added on resume training scenario. 
Ex: 
if not any(isinstance(cb, KubeflowProgressCallback) for cb in self.callback_handler.callbacks):
    callback = KubeflowProgressCallback(metrics_port)
    self.add_callback(callback)

robert-bell replied:
I think the current `if` statement will always evaluate to true - the callback will never be in the handlers because we've only just created a new instance on the line above.

> Maybe we should do a callback type check here rather than instance check to avoid the KubeflowProgressCallback being added on resume training scenario.

@efazal is the callback stored in the checkpoint? If so, how? I'm thinking, if it's a pickle it might be a problem because local/nested classes aren't pickleable.

efazal replied:
Yeah true, probably the existing check is sufficient. As you said, there possibly can't be another instance of progress callback. I was more of so saying, even if there was one, a type check would avoid it. But its a nit pick I suppose. So i'm okay with it as is even.

Also regarding callback stored in checkpoint - no it's not stored in the checkpoint.

efazal commented on kubeflow/trainer/rhai/transformers.py:
```diff
@@ -22,15 +54,328 @@ class TransformersTrainer:
+    instrumentation_code = inspect.getsource(_create_progression_instrumentation)
+    instrumentation_code = textwrap.dedent(instrumentation_code)
+
+    lines = instrumentation_code.split("\n")
```
Line 272-292: There's pattern matching going on to identify the start and end of the injection code based on "def" and "return" statements, but probably a better approach would suit? since this might break if someone were to add to new helper function in "_create_progression_instrumentation()"  at the start with a return statement and code injected will be incorrect.

abhijeet-dhumal replied:
Great catch 🙌 
You're absolutely right - this pattern matching is fragile. If someone adds a helper function with a return statement inside _create_progression_instrumentation(), it might  break.
What if I use the title for begin and end points : 
`BEGIN/END_PROGRESSION_INSTRUMENTATION_CODE` ?

efazal replied:
I would suggest a different approach altogether. How about the instrumentation code in a separate file. Separate the instrumentation logic or injection code and the invocation code. So that way you can construct the code to be injected piece by piece by importing using `inspect.get_source()`. I've added some pseudo code below. I hope it makes sense. 

```suggestion
# New file for instrumentation code - progression_intrumentation_code.py
Class ProgressionInstrumentation()
        # keep imports at class level and not module level so the injected code doesn't run into import errors
        __innit__(self, metric_port):
      
        @dataclass
        class ProgressionMetricsState:
        
        class KubeflowProgressCallback(TrainerCallback):
      
        # rest of the intrumentation code within this ProgressionInstrumentation() class. And feel free to break in many classes or func's as you like if required
   
def apply_progression_tracking():
     _original_init = trainer_module.Trainer.__init__
     def _instrumented_trainer_init(self, *args, **kwargs):      
     # Doesn't need to return anything
     
     
# In the transformer.py - you need to construct the injection code

# Replace get_transformers_instrumentation_wrapper() 
def get_progression_injection_code(metrics_port) -> str:

    import progression_intrumentation_code as progress

    intrumentation_code = inspect.getsource(progress.ProgressionInstrumentation)
    patch_code = inspect.getsource(progress.apply_progression_tracking)
    monkey_patch_invocation = textwrap.dedent(
        try:
            apply_progression_tracking()  # applies your callsbacks
        except:
            raise
    ).strip()
    
    # construct injection code
    complete_code = (
        f"{intrumentation_code}\n\n\n"
        f"{patch_code}\n\n\n"
        f"{monkey_patch_invocation}\n"
    )
    
    return complete_code
```

abhijeet-dhumal replied:
I have refactored the code based on the recent format suggested by Rob here : https://github.com/opendatahub-io/kubeflow-sdk/pull/21#discussion_r2524129705
Please check and let me know if we still need the separate instrumentation based files ..

robert-bell replied:
I might be missing something, but I don't follow why we need to parse the lines? Is there any reason we can't do this?

```py
instrumentation_code = inspect.getsource(_create_progression_instrumentation)
instrumentation_code = textwrap.dedent(instrumentation_code)

wrapper = f"""

{instrumentation_code}
apply_progression_tracking , *_ = {_create_progression_instrumentation.__name__}(metrics_port={metrics_port})
apply_progression_tracking()
"""

# add user code here...
```

efazal replied:
Yeah my suggestion is similar to what Rob had proposed. A separate file for the instrumentation code is better I thought since I'll be adding checkpointing instrumentation code as well. It'll be clean and maintainable if we have the instrumentation code separate. I don't have strong opinion on that tho. 

And good question about the parsing, why are are we doing it even?

robert-bell replied:
@efazal I was thinking we can keep your instrumentation code completely separate so it can be tested in isolation - something like this

```py
progress_instrumentation_code = textwrap.dedent(inspect.getsource(_create_progression_instrumentation))
checkpoint_instrumentation_code = textwrap.dedent(inspect.getsource(_create_checkpoint_instrumentation))

wrapper = f"""
{progress_instrumentation_code}
apply_progression_tracking, *_ = {_create_progression_instrumentation.__name__}(metrics_port={metrics_port})
apply_progression_tracking()

{checkpoint_instrumentation_code}
apply_checkpointing,  *_ = {_create_checkpoint_instrumentation.__name__}()
apply_checkpointing()

# add user code here...
"""
```

abhijeet-dhumal replied:
@robert-bell @efazal  I have refactored the implementation accordingly yesterday, I see the current implementation is perfectly compatible for above format.. 
Please review!

robert-bell commented on kubeflow/trainer/backends/kubernetes/backend.py:
```diff
@@ -215,7 +220,9 @@ def train(
             pod_template_overrides = spec_section.get("podTemplateOverrides")
 
-        # Generate unique name for the TrainJob if not provided
+        if isinstance(trainer, (TransformersTrainer, TrainingHubTrainer)):
```
nit:

```suggestion
        if isinstance(trainer, RHAITrainer):
```

briangallagher replied:
nit: I'd probably lean towards excluding TrainingHubTrainer until the progress logic is in place for the Trainer

abhijeet-dhumal replied:
Currently the problem I'm seeing is : In Python 3.9, you cannot use isinstance() with Union types
That's why 3.9 based unit tests are failing 😢

robert-bell replied:
@briangallagher are we supporting py312 only? That's the version that's going to be in the universal images. Are you planning on making a PR to update the version?

abhijeet-dhumal replied:
Now I have added small tweak here, to keep using RHAITrainer and still support Python 3.9 based unit tests using `get_args`

briangallagher replied:
yes, we only support 3.12 due to the APICC/Universal image constraint.
However, I didn't modify the code base in order to keep more in sync with upstream and minimise differences. 
I'd probably continue with that approach i.e. inline with upstream, but happy to go with the majority. 
Any thoughts @kramaranya @MStokluska @Fiona-Waters @szaher

kramaranya replied:
Yeah, I'd lean towards aligning with upstream for now to avoid merge conflicts

abhijeet-dhumal replied:
current approach used here is compatible for python 3.9 to 3.12 ✅

efazal commented on kubeflow/trainer/rhai/transformers.py:
```diff
@@ -22,15 +54,328 @@ class TransformersTrainer:
+
+    if trainer.enable_progression_tracking:
+        wrapper_code = get_transformers_instrumentation_wrapper(
+            metrics_port=trainer.metrics_port,
```
Since the user has the option to provide metrics port and poll interval, maybe add validations to ensure the user doesn't provide incorrect port number (reserved ports, negative numbers, etc..) and same for poll interval. 

Maybe add validation post initialization of transformer trainer to raise any invalid inputs related error

abhijeet-dhumal replied:
I have added the validations and test cases for these scenarios, please review!

efazal commented on kubeflow/trainer/rhai/transformers.py:
```diff
@@ -22,15 +54,328 @@ class TransformersTrainer:
+            self.start_time = time.time()
+
+            if state.is_world_process_zero and self.server is None:
+                server = http.server.HTTPServer(
```
For robustness, maybe some error handling here. If the user provided port is already in use, we could fallback to the sdk defined default port instead of crashing right way?

robert-bell replied:
unfortunately I think we'd just have to fail here - we can't easily use a default fallback because the control plane would need to know that the port had changed.

efazal replied:
Ah fair point. Perhaps, just graceful error handling here would be nice for user experience I suppose?

abhijeet-dhumal replied:
I have added the graceful error handling now, Please review!

robert-bell commented on kubeflow/trainer/rhai/utils.py:
```diff
@@ -30,8 +29,52 @@ def get_trainer_cr_from_rhai_trainer(
-    raise ValueError(f"Unknown trainer {trainer}.")
+    if spec_annotations is None:
+        return progression_annotations
+    return {**spec_annotations, **progression_annotations}
```
nit: I'd suggest merging these the other way round so users can override the progression annotations if they want. They shouldn't need to, but it may be surprising if they can't.

```suggestion
    # merge spec_annotations last to allow users to override the progress annotations
    return {**progression_annotations, **spec_annotations}
```

abhijeet-dhumal replied:
here do you mean to allow user-provided annotations should have precedence ?

efazal commented on kubeflow/trainer/rhai/transformers.py:
```diff
@@ -22,15 +54,328 @@ class TransformersTrainer:
+                        metric_value = value
+                    else:
+                        try:
+                            import torch
```
nit pick: not sure if importing torch multiple times is a good thing.

abhijeet-dhumal replied:
Now there's only one torch import in the entire file , It's a lazy import inside the on_log method instead of at the module level , 
Currently for TransformersTrainer, torch is always available, so we could move it to line 154-161. 
But to be on safer side as it adds negligible overhead.

abhijeet-dhumal replied:
I'd be happy to discuss if I've misunderstood you!

efazal replied:
No I was thinking it was better to move at method level, but yeah, this won't have a performance impact. So it doesn't really matter. I'm even okay if we decide to leave it as is. Just a nit anyway.

efazal reviewed (commented):
A few comments, but otherwise looks good. Thanks!

kramaranya commented on kubeflow/trainer/rhai/__init__.py:
```diff
@@ -1,13 +1,35 @@
+# Copyright 2024 The Kubeflow Authors.
```
```suggestion
# Copyright 2025 The Kubeflow Authors.
```

kramaranya commented on kubeflow/trainer/rhai/constants.py:
```diff
@@ -0,0 +1,23 @@
+ANNOTATION_METRICS_PORT = "trainer.opendatahub.io/metrics-port"
+ANNOTATION_METRICS_POLL_INTERVAL = "trainer.opendatahub.io/metrics-poll-interval"
+ANNOTATION_FRAMEWORK = "trainer.opendatahub.io/framework"
+ANNOTATION_TRAINER_STATUS = "trainer.opendatahub.io/trainerStatus"
```
do we need this? doesn't look like we use it in the sdk

abhijeet-dhumal replied:
Actually this one is just to make sdk aware that the metrics will be stored in this named annotation in the TrainJob CR..
I was planning to add helping method `get_job_progress` for user to get realtime progress in notebook via SDK similar to `get_job_logs`
But I will add it in follow up PR..

abhijeet-dhumal replied:
I'd be happy to discuss if I've misunderstood you!

kramaranya commented on kubeflow/trainer/rhai/transformers.py:
```diff
@@ -22,15 +54,328 @@ class TransformersTrainer:
     resources_per_node: Optional[dict] = None
     env: Optional[dict[str, str]] = None
 
-    enable_jit_checkpointing: bool = True
```
is the plan to still have this arg but add it in the follow-up pr?

abhijeet-dhumal replied:
Yup, that's right! 🙌 
it doesn't makes sense at this moment as this config has no implementation set to it yet

abhijeet-dhumal replied:
cc: @efazal

kramaranya commented on kubeflow/trainer/rhai/transformers.py:
```diff
@@ -22,15 +54,328 @@ class TransformersTrainer:
+
+        wrapped_code = wrapper_code.replace("{{user_func_import_and_call}}", "{func_code}")
+        trainer_crd.command = [
+            cmd.replace("{func_code}", wrapped_code) if "{func_code}" in cmd else cmd
```
Since we already format func_code in [get_command_using_train_func](https://github.com/opendatahub-io/kubeflow-sdk/blob/90e5d7e5002db190d1a565b41069ce73219c6037/kubeflow/trainer/backends/kubernetes/utils.py#L337-L340) before, is there any func_code left in trainer_crd.command to perform this?

abhijeet-dhumal replied:
I have refactored the execution flow, please review and let me know if this concern still persists !

kramaranya commented on kubeflow/trainer/rhai/transformers_test.py:
```diff
@@ -0,0 +1,361 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tests for TransformersTrainer and instrumentation wrapper generation."""
```
Can we also test get_trainer_cr_from_transformers_trainer?

abhijeet-dhumal replied:
I have added the tests, Please review!

abhijeet-dhumal committed:
fix: Minor bug fixes

Signed-off-by: abhijeet-dhumal <abhijeetdhumal652@gmail.com>

abhijeet-dhumal committed:
fix: Add test for get_trainer_cr_from_transformers_trainer

Signed-off-by: abhijeet-dhumal <abhijeetdhumal652@gmail.com>

abhijeet-dhumal committed:
fix: add gracefull error handling

Signed-off-by: abhijeet-dhumal <abhijeetdhumal652@gmail.com>

abhijeet-dhumal committed:
fix: resolve python 3.9 based errors

Signed-off-by: abhijeet-dhumal <abhijeetdhumal652@gmail.com>

abhijeet-dhumal committed:
fix: Refactor wrapper injection flow

Signed-off-by: abhijeet-dhumal <abhijeetdhumal652@gmail.com>

abhijeet-dhumal committed:
fix: Minor fixes

Signed-off-by: abhijeet-dhumal <abhijeetdhumal652@gmail.com>

abhijeet-dhumal committed:
fix: minor fix

Signed-off-by: abhijeet-dhumal <abhijeetdhumal652@gmail.com>

abhijeet-dhumal committed:
fix: refine transformerTrainer tests

Signed-off-by: abhijeet-dhumal <abhijeetdhumal652@gmail.com>

abhijeet-dhumal committed:
fix: adjust callback logic and calculations

Signed-off-by: abhijeet-dhumal <abhijeetdhumal652@gmail.com>

abhijeet-dhumal committed:
fix: minor formatting fix

Signed-off-by: abhijeet-dhumal <abhijeetdhumal652@gmail.com>

abhijeet-dhumal committed:
fix: round of epoch count to get accurate results

Signed-off-by: abhijeet-dhumal <abhijeetdhumal652@gmail.com>

abhijeet-dhumal committed:
fix: minor linting error

Signed-off-by: abhijeet-dhumal <abhijeetdhumal652@gmail.com>

robert-bell commented on kubeflow/trainer/rhai/transformers.py:
```diff
@@ -22,15 +63,452 @@ class TransformersTrainer:
+            self.send_response(200)
+            self.send_header("Content-Type", "application/json")
+            self.end_headers()
+            self.wfile.write(_get_progression_metrics_json().encode("utf-8"))
```
sorry - one small suggestion here for better error reporting


```suggestion
            try: 
                payload = _get_progression_metrics_json()
            except Exception e:
                print(f"[Kubeflow] Failed to create progress metrics payload: {e}")
                self.send_error(500)
            else:
                self.send_response(200)
                self.send_header("Content-Type", "application/json")
                self.end_headers()
                self.wfile.write(payload.encode("utf-8"))
```

abhijeet-dhumal replied:
thanks @robert-bell !
Let me update it right away 👀

abhijeet-dhumal replied:
Done ✅

robert-bell commented:
lgtm. Thanks @abhijeet-dhumal 

Let's get approval from the devx team then we can get it merged!

abhijeet-dhumal committed:
fix: add error handling to progress metrics HTTP endpoint

Signed-off-by: abhijeet-dhumal <abhijeetdhumal652@gmail.com>

abhijeet-dhumal commented:
@briangallagher @szaher @Fiona-Waters @kramaranya 
The PR is up and ready for review !
Please share your approval if the feature added seems legit and share any suggestions on my way if any aspect I might have missed.. 
Thanks!

abhijeet-dhumal committed:
fix: remove framework annotation

Signed-off-by: abhijeet-dhumal <abhijeetdhumal652@gmail.com>

abhijeet-dhumal committed:
fix: refine callback logic and calculations for the precise training progression metrics

Signed-off-by: abhijeet-dhumal <abhijeetdhumal652@gmail.com>

kramaranya reviewed (commented):
Thanks for this amazing work, @abhijeet-dhumal!
/lgtm

briangallagher commented:
great work @abhijeet-dhumal 
/lgtm
